editor/plugins/dedicated_server_export_plugin.cpp

```diff
@@ -0,0 +1,126 @@
+	Ref<EditorExportPreset> preset = get_export_preset();
+	ERR_FAIL_COND_V(preset.is_null(), false);
+
+	bool valid_prop = false;
+	bool is_server = preset->get(EXPORT_OPTION_IS_DEDICATED_SERVER, &valid_prop);
+
+	if (valid_prop) {
+		return is_server;
+	}
+
+	return false;
+}
+
+PackedStringArray DedicatedServerExportPlugin::_get_export_features(const Ref<EditorExportPlatform> &p_platform, bool p_debug) const {
+	PackedStringArray ret;
+	if (is_dedicated_server()) {
+		ret.append("dedicated_server");
+	}
+	return ret;
+}
+
+uint64_t DedicatedServerExportPlugin::_get_customization_configuration_hash() const {
+	return (uint64_t)is_dedicated_server();
+}
+
+bool DedicatedServerExportPlugin::_begin_customize_resources(const Ref<EditorExportPlatform> &p_platform, const Vector<String> &p_features) const {
+	return is_dedicated_server();
+}
+
+Ref<Resource> DedicatedServerExportPlugin::_customize_resource(const Ref<Resource> &p_resource, const String &p_path) {
+	if (p_resource->get_dedicated_server_export_type() == Resource::DEDICATED_SERVER_EXPORT_KEEP) {
+		return Ref<Resource>();
+	}
+
+	if (const Texture2D *texture = Object::cast_to<Texture2D>(p_resource.ptr())) {
+		Ref<PlaceholderTexture2D> placeholder;
+		placeholder.instantiate();
+		placeholder->set_size(texture->get_size());
+		return placeholder;
+	}
+
+	if (const Texture2DArray *texture = Object::cast_to<Texture2DArray>(p_resource.ptr())) {
+		Ref<PlaceholderTexture2DArray> placeholder;
+		placeholder.instantiate();
+		placeholder->set_size(Size2i(texture->get_width(), texture->get_height()));
+		placeholder->set_layers(texture->get_layers());
+		return placeholder;
+	}
+
+	if (const Texture3D *texture = Object::cast_to<Texture3D>(p_resource.ptr())) {
+		Ref<PlaceholderTexture3D> placeholder;
+		placeholder.instantiate();
+		placeholder->set_size(Vector3i(texture->get_width(), texture->get_height(), texture->get_depth()));
+		return placeholder;
+	}
+
+	if (const Cubemap *cubemap = Object::cast_to<Cubemap>(p_resource.ptr())) {
+		Ref<PlaceholderCubemap> placeholder;
+		placeholder.instantiate();
+		placeholder->set_size(Size2i(cubemap->get_width(), cubemap->get_height()));
+		placeholder->set_layers(cubemap->get_layers());
+		return placeholder;
+	}
+
+	if (const CubemapArray *cubemap = Object::cast_to<CubemapArray>(p_resource.ptr())) {
+		Ref<PlaceholderCubemapArray> placeholder;
+		placeholder.instantiate();
+		placeholder->set_size(Size2i(cubemap->get_width(), cubemap->get_height()));
+		placeholder->set_layers(cubemap->get_layers());
+		return placeholder;
+	}
+
+	if (Object::cast_to<Material>(p_resource.ptr()) != nullptr) {
+		Ref<PlaceholderMaterial> placeholder;
+		placeholder.instantiate();
+		return placeholder;
+	}
+
+	if (const Mesh *mesh = Object::cast_to<Mesh>(p_resource.ptr())) {
+		Ref<PlaceholderMesh> placeholder;
+		placeholder.instantiate();
+		placeholder->set_aabb(mesh->get_aabb());
+		return placeholder;
+	}
```

Comment:
The logic here, (instantiating a placeholder resource which derives from a source resource), I don't think this is the best place for it ( `dedicated_server_export_plugin.cpp`).  It seems like each class itself should be responsible for understanding how to create a placeholder version of itself...  Perhaps each of these resources should have a `to_placeholder()` or similar function, and `dedicated_server_export_plugin.cpp` can be consumer of those api's instead of `dedicated_server_export_plugin.cpp` requiring knowledge of these details and being the implementer for each type?

Reply:
This is a great idea! That would certainly make it easier to progressively add more placeholders. I'll work this into the next revision.